crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -683,6 +693,79 @@ impl EventCache {
             }
         }
     }
+
+    #[cfg(feature = "experimental-search")]
+    fn parse_timeline_event(event: &TimelineEvent) -> Option<AnySyncMessageLikeEvent> {
+        use ruma::events::AnySyncTimelineEvent;
+
+        if event.kind.is_utd() {
+            return None;
+        }
+
+        match event.raw().deserialize() {
+            Ok(event) => match event {
+                AnySyncTimelineEvent::MessageLike(event) => Some(event),
+                AnySyncTimelineEvent::State(_) => None,
+            },
+
+            Err(e) => {
+                warn!("failed to parse event: {e:?}");
+                None
+            }
+        }
+    }
+
+    /// Takes a [`TimelineEvent`] and passes it to the [`RoomIndex`] of the
+    /// given room which will add/remove/edit an event in the index based on
+    /// the event type.
+    #[cfg(feature = "experimental-search")]
+    async fn search_indexing_task(
+        client: WeakClient,
+        linked_chunk_update_sender: Sender<RoomEventCacheLinkedChunkUpdate>,
+    ) {
+        let Some(client) = client.get() else {
+            trace!("Client is shutting down, not spawning thread subscriber task");
+            return;
+        };
+
+        let mut linked_chunk_update_receiver = linked_chunk_update_sender.subscribe();
+
+        loop {
+            match linked_chunk_update_receiver.recv().await {
+                Ok(room_ec_lc_update) => {
+                    let OwnedLinkedChunkId::Room(room_id) =
+                        room_ec_lc_update.linked_chunk_id.clone()
+                    else {
+                        trace!("Received non-room updates, ignoring.");
+                        continue;
+                    };
+
+                    let timeline_events = room_ec_lc_update.events();
+
+                    for event in timeline_events {
+                        if let Some(message_event) = Self::parse_timeline_event(&event) {
+                            let _ = client
+                                .search_index()
+                                .lock()
+                                .await
```

Comment:
To help perf, you can lock the `search_index` only once, above the loop, if `!timeline_events.is_empty()`, to avoid locking/unlocking in a loop. This is called "loop invariant code motion", in compiler engineering; hoist the "constant" parts of a loop on top of it, to avoid repeating the same work.